std/traits.d

```diff
@@ -2658,6 +2659,142 @@ template RepresentationTypeTuple(T)
+and pointers, the layout is $(D AliasSeq!(size_t(0), T)).)
+
+$(LI If `T` is a `struct` or `union` type, then `Layout!T` is an `AliasSeq`
+containing each field preceded by its offset. Fields that are in turn of
+`struct` or `union` types are decomposed transitively. The result is ordered
+in increasing offset order. (Two or more fields may have the same offset due to
+`union`s.) For a given offset, types are ordered lexicographically by their
+name.)
+
+$(LI If `T` is a `class` or `interface` type, the layout is similar to the one
+return for `struct`s, with one difference: classes and interfaces have
+additional opaque data managed by the compiler. Usually these are placed at the
+front of the object. This data is represented in the layout as pointers to
+opaque types. The presence, number, and types of additional fields are
+implementation-dependent and should not be relied upon in portable code. All
+these additional types are modeled as $(D __ClassMeta!(T, index)) where `T` is
+the class or interface type, and `index` is a sequence number starting at `0`.
+In the current implementations, the first field in the layout has type
+$(D __ClassMeta!(T, 0)*) and contains a pointer to the virtual functions table
+associated with `T`. The second field, present in `class` objects but not in
+`interface` objects, has type $(D __ClassMeta!(T, 1)*) and represents a monitor
+type used for synchronization. These details may change in the future.)
+)
+*/
+template Layout(T)
+{
+    private template Add(size_t offset, U...)
+    {
+        static if (!U.length)
+            alias Add = AliasSeq!();
+        else
+            alias Add = AliasSeq!(U[0] + offset, U[1],
+                                     Add!(offset, U[2 .. $]));
+    }
+
+    static if (!is(T == struct) && !is(T == union)
+                    && !is(T == class) && !is(T == interface))
+    {
+        alias Layout = AliasSeq!(size_t(0), T);
+    }
+    else
+    {
+        // Simple mergesort by offset implementation
+        private template Sort(T...)
+        {
+            template Merge(T...)
+            {
+                template With(U...)
+                {
+                    static if (T.length == 0)
+                        alias With = U;
+                    else static if (U.length == 0)
+                        alias With = T;
+                    else static if (T[0] < U[0]
+                            || T[0] == U[0] && T[1].stringof <= U[1].stringof)
+                        alias With =
+                            AliasSeq!(T[0], T[1], Merge!(T[2 .. $]).With!U);
+                    else
+                        alias With =
+                            AliasSeq!(U[0], U[1], Merge!T.With!(U[2 .. $]));
+                }
+            }
+            static if (T.length <= 2)
+                alias Sort = T;
+            else
+                alias Sort = Merge!(Sort!(T[0 .. T.length / 2]))
+                    .With!(Sort!(T[T.length / 2 .. $]));
```

Comment:
If T.length = 6 (e.g. in `Layout!{int a, b, c;}` then is the merge doing `Merge!(0UL, int, 4UL).With!(int, 8UL, int)`? Should maybe check if T.length / 2 == odd or even and deal appropriately?